### PR TITLE
Proof best practices: autoloads_versus_internal_nodes

### PR DESCRIPTION
**Pull request by NathanLovato**
I mostly simplified the text, moved some notes around, also used double backticks for inline code. Mostly proofing/editing work.

### CODE REVIEW

File: getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst [hunk 1]
```diff
@@ -1,103 +1,95 @@
...
+effects comes in. Say we call that class ``Sound``, you can use it from anywhere
+in your project by calling ``Sound.play("coin_pickup.ogg")``. This solves the
+problem in the short term but causes more problems:
+
+1. **Global state**: one object is now responsible for all objects' data. If the
+   ``Sound`` class has errors or doesn't have an AudioStreamPlayer available,
+   all the nodes calling it can break.
+
+2. **Global access**: now that any object can call ``Sound.play(sound_path)``
+   from anywhere, there's no longer an easy way to find the source of a bug.
```
Comment (willnationsdev) [inline getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst]: > there's no longer an easy way to find the source of a bug.

I think that we should elaborate here on what we mean. *How* does this make it harder to find bugs? What could potentially happen and what could we have trouble doing in our attempt to solve it?

File: getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst [hunk 2]
```diff
@@ -1,103 +1,95 @@
...
+Another reason to use an Autoload can be that you want to reuse the same method
+or data across many scenes.
+
+In the case of functions, you can create a new type of ``Node`` that provides
+that feature for an individual scene using the :ref:`class_name
+<doc_scripting_continued_class_name>` keyword in GDScript.
+
+When it comes to data, you can either:
+
+1. Create a new type of :ref:`Resource <class_Resource>` to share the data
```
Comment (willnationsdev) [inline getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst]: Period at the end of this sentence.

File: getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst [hunk 3]
```diff
@@ -1,103 +1,95 @@
...
+Managing shared functionality or data
+=====================================
+
+Another reason to use an Autoload can be that you want to reuse the same method
+or data across many scenes.
+
+In the case of functions, you can create a new type of ``Node`` that provides
+that feature for an individual scene using the :ref:`class_name
+<doc_scripting_continued_class_name>` keyword in GDScript.
+
```
Comment (willnationsdev) [inline getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst]: Another thing that is important to discuss here for "in the case of functions" is the difference between pure and impure functions and how one can create a library of independently testable static functions within script classes. I elaborate more on this content in [section `1.b.i` of this reddit comment](https://www.reddit.com/r/godot/comments/dhjohq/best_ways_to_implement_systems_with_many/f3om27g/). If you would like, you could add a tidbit here in a second paragraph, before the "when it comes to data" section.

File: getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst [hunk 4]
```diff
@@ -1,103 +1,95 @@
...
+- **Systems with a wide scope**: If the singleton is managing its own
+  information and not invading the data of other objects, then it's a great way to
+  create systems that handle broad-scoped tasks. For example, a quest or a
+  dialogue system.
+
+Until Godot 3.1, another use was just for convenience: autoloads have a global
+variable for their name generated in GDScript, allowing you to call them from
+any script file in your project. But now, you can use the ``class_name`` keyword
+instead to get autocompletion for a type in your entire project.
```
Comment (willnationsdev) [inline getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst]: I think this section is more useful if we actually include an example of using a singleton/script class as a namespace.

File: getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst [hunk 5]
```diff
@@ -1,103 +1,95 @@
...
+
+.. note::
+
+   Autoload is not exactly a Singleton. Nothing prevents you from instancing
+   copies of an auto-loaded node. It is only a tool that makes a node load
+   automatically at the root of your scene tree, regardless of your game's node
```
Comment (willnationsdev) [inline getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst]: Correction: "that makes a node load automatically \[as a child of\] the root of your scene tree..." I would recommend explicitly telling people that it will exist at `/root/<autoload_name>`. Clarity is key here.

File: getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst [hunk 6]
```diff
@@ -1,103 +1,95 @@
...
+   structure or the scene you try.
```
Comment (willnationsdev) [inline getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst]: I was confused by what was meant with "the scene you try" here. Maybe rephrase it? Did you mean the scene one tries to `change_scene` to?

Comment (NathanLovato) [reply getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst]: I meant the autoload will be there even when you press F6 to run an individual node or scene that is not your app's main scene

File: getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst (hunk 1 above)
Comment (NathanLovato) [reply getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst]: I took this from the original article. I wouldn't use a singleton for things like these so I don't have examples from experience but I'd guess that:

1. You have to navigate to the Sound class to change the code
2. If you run out of streams and the Sound class cycles and e.g. replaces a long audio sample with a new one, it's not that easy to figure out why that's happening in the middle of a project, especially as you won't get a backtrace and when you try the scene in isolation with F6, everything may work fine.

Does that seem good enough of an example? In my exp. the issues of singletons are not obvious at first, they really start to hurt when your project grows and you have a hard dependency in many places at once. It's difficult to give an example of that.

File: getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst (hunk 3 above)
Comment (NathanLovato) [reply getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst]: Yup definitely, that's actually mostly how we use them - function libraries.

**Comment by NathanLovato** [comment]
Thanks, it's getting late here, I'll get to your feedback tomorrow!

File: getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst (hunk 1 above)
Comment (willnationsdev) [reply getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst]: Well, those examples work well too. But my primary concern with things like this is that the location of a problem, namely the context in which the method is called, is no longer clear when the functionality can be accessed from anywhere. Rather than dealing with a subset of nodes internal to a scene, your potential domain of trouble-making code spans the entire project. Any code anywhere could have passed erroneous data to the `Sound` singleton. When code is kept inside a scene, there is no realy concern for this problem as only a handful of scripts are usually involved total, let alone dealing with the topic that the singleton handles.

File: getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst [hunk 7]
```diff
@@ -1,103 +1,95 @@
...
+  dialogue system.
+
+Until Godot 3.1, another use was just for convenience: autoloads have a global
+variable for their name generated in GDScript, allowing you to call them from
+any script file in your project. But now, you can use the ``class_name`` keyword
+instead to get autocompletion for a type in your entire project.
+
+.. note::
+
+   Autoload is not exactly a Singleton. Nothing prevents you from instancing
```
Comment (willnationsdev) [inline getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst]: > instancing

Do we want to say `instantiating` here? I think we should.

File: getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst (hunk 6 above)
Comment (willnationsdev) [reply getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst]: Ah, in that case, I would recommend converting `or the scene you try` to `or which scene you run`.

File: getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst (hunk 3 above)
Comment (NathanLovato) [reply getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst]: Thinking of this again, I don't think this discussion is a good fit for a guide about "when and how to use auto-loaded nodes in Godot" because you can use a gdscript file that doesn't extend anything to create a library of static functions, i.e. with class_name

File: getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst [hunk 8]
```diff
@@ -1,103 +1,109 @@
 .. _doc_autoloads_versus_internal_nodes:
 
-Autoloads versus internal nodes
+Autoloads versus regular nodes
+==============================
+
+Godot offers a feature to automatically load nodes at the root of your project,
+allowing you to access them globally, that can fulfill the role of a Singleton:
+:ref:`doc_singletons_autoload`. These auto-loaded nodes are not freed when you
+change the scene from code with ``SceneTree.change_scene``.
```
Comment (willnationsdev) [inline getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst]: Should we replace this with a direct link to the method in the API docs?

    change the scene from code with :ref:`SceneTree.change_scene <class_SceneTree_method_change_scene>`.

**Comment by akien-mga** [comment]
What's the status on this PR?

**Comment by NathanLovato** [comment]
I corrected the typos. The other notes from Will involve writing some new paragraphs, which I would rather do or have someone do in a separate PR.

But I transferred my docs fork some time ago and it seems GitHub lost track of it as a result. I'll open a new PR that addresses the typos Will pointed out.
